### PR TITLE
[commands][types] Type hint commands-ext

### PR DESCRIPTION
**Pull request by bijij**
## Summary

This PR attempts to type-hint the commands extension using ParamSpec.

Completed:
 - `bot`
 - `cog.py`
 -  `core.py`
 -  `context.py`
 - ~~`errors`~~ (#7382)
 - ~~`cooldowns`~~  (#7339)
 
 To-Do;
  - `help` ? -- this is getting a serious rework?
  - `converters` 
  - `view`

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/bot.py [hunk 1]
```diff
@@ -129,10 +148,10 @@ def __init__(self, command_prefix, help_command=_default, description=None, **op
     # internal helpers
 
     def dispatch(self, event_name, *args, **kwargs):
-        super().dispatch(event_name, *args, **kwargs)
+        super().dispatch(event_name, *args, **kwargs)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/bot.py:151]: any attribute of `Client` accessed in `BotBase` will need to be type ignored. unless we do
`assert isinstance(self, Client)` at the top of each call or have BotBase subclass Client.

File: discord/ext/commands/bot.py [hunk 2]
```diff
@@ -795,26 +814,26 @@ def reload_extension(self, name, *, package=None):
             # if the load failed, the remnants should have been
             # cleaned from the load_extension function call
             # so let's load it from our old compiled library.
-            lib.setup(self)
+            lib.setup(self)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/bot.py:817]: Unless we add a HasSetup protocol, we'll need to type-ignore `setup` or do something simular to what is done in `_call_module_finalizers`

File: discord/ext/commands/bot.py [hunk 3]
```diff
@@ -938,11 +957,11 @@ class be provided, it must be similar enough to :class:`.Context`\'s
 
         invoker = view.get_word()
         ctx.invoked_with = invoker
-        ctx.prefix = invoked_prefix
+        ctx.prefix = invoked_prefix  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/bot.py:960]: Function would need rework to remove this, as `invoked_prefix` type changes from `List[str] | str` to `str` throughout function.

File: discord/ext/commands/bot.py [hunk 4]
```diff
@@ -645,13 +664,13 @@ def _load_from_module_spec(self, spec, key):
         else:
             self.__extensions[key] = lib
 
-    def _resolve_name(self, name, package):
+    def _resolve_name(self, name: str, package: Optional[str]) -> str:
         try:
             return importlib.util.resolve_name(name, package)
```
Comment (bijij) [inline discord/ext/commands/bot.py:669]: This doesn't type-check currently due to a typeshed issue. python/typeshed#5730

File: discord/ext/commands/bot.py [hunk 5]
```diff
@@ -39,21 +45,34 @@
...
     'when_mentioned',
     'when_mentioned_or',
     'Bot',
     'AutoShardedBot',
 )
 
-def when_mentioned(bot, msg):
+T = TypeVar('T')
+CFT = TypeVar('CFT', bound='CoroFunc')
+CXT = TypeVar('CXT', bound='Context')
```
Comment (bijij) [inline discord/ext/commands/bot.py:66]: Do we want a proper naming convention set for typevars?

File: discord/ext/commands/cog.py [hunk 6]
```diff
@@ -165,8 +184,12 @@ class Cog(metaclass=CogMeta):
     When inheriting from this class, the options shown in :class:`CogMeta`
     are equally valid here.
     """
+    __cog_name__: ClassVar[str]
+    __cog_settings__: ClassVar[Dict[str, Any]]
+    __cog_commands__: ClassVar[List[Command]]
+    __cog_listeners__: ClassVar[List[Tuple[str, str]]]
```
Comment (bijij) [inline discord/ext/commands/cog.py:190]: I'm not sure if I can get the vars defined in the metaclass to carrry down to the class without doing this.
Ideas?

File: discord/ext/commands/cog.py [hunk 7]
```diff
@@ -190,12 +213,12 @@ def __new__(cls, *args, **kwargs):
                 parent = lookup[parent.qualified_name]
 
                 # Update our parent's reference to our self
-                parent.remove_command(command.name)
-                parent.add_command(command)
+                parent.remove_command(command.name)  # type: ignore
+                parent.add_command(command)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/cog.py:217]: See comment on Command.parent

File: discord/ext/commands/context.py [hunk 8]
```diff
@@ -86,23 +121,38 @@ class Context(discord.abc.Messageable):
...
+        args: List[Any] = MISSING,
+        kwargs: Dict[str, Any] = MISSING,
+        prefix: Optional[str] = None,
+        command: Optional[Command] = None,
+        invoked_with: Optional[str] = None,
+        invoked_parents: List[str] = MISSING,
+        invoked_subcommand: Optional[Command] = None,
+        subcommand_passed: Optional[str] = None,
+        command_failed: bool = False,
+        current_parameter: Optional[inspect.Parameter] = None,
```
Comment (bijij) [inline discord/ext/commands/context.py:138]: `attrs` was expanded to help the type-checker, undo this?

File: discord/ext/commands/context.py [hunk 9]
```diff
@@ -133,17 +183,9 @@ async def invoke(self, command, /, *args, **kwargs):
...
-        arguments = []
-        if command.cog is not None:
-            arguments.append(command.cog)
-
-        arguments.append(self)
-        arguments.extend(args)
+        return await command(self, *args, **kwargs)
 
-        ret = await command.callback(*arguments, **kwargs)
-        return ret
```
Comment (bijij) [inline discord/ext/commands/context.py:144]: Simplfies the function, would you consider this outside the scope?

File: discord/ext/commands/context.py [hunk 10]
```diff
@@ -187,7 +229,7 @@ async def reinvoke(self, *, call_hooks: bool = False, restart: bool = True):
 
         if restart:
             to_call = cmd.root_parent or cmd
-            view.index = len(self.prefix)
+            view.index = len(self.prefix or '')
```
Comment (bijij) [inline discord/ext/commands/context.py:232]: `prefix` could be `None` on an invalid context.

File: discord/ext/commands/context.py [hunk 11]
```diff
@@ -248,26 +293,26 @@ def channel(self):
...
         return self.message.author
 
     @discord.utils.cached_property
-    def me(self):
+    def me(self) -> Union[Member, ClientUser]:
         """Union[:class:`.Member`, :class:`.ClientUser`]:
         Similar to :attr:`.Guild.me` except it may return the :class:`.ClientUser` in private message contexts.
         """
-        return self.guild.me if self.guild is not None else self.bot.user
+        return self.guild.me if self.guild is not None else self.bot.user  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/context.py:307]: Type-checker thinks this could be None, which is unlikely here.

File: discord/ext/commands/context.py [hunk 12]
```diff
@@ -326,12 +371,12 @@ async def send_help(self, *args):
...
 
         try:
-            entity.qualified_name
+            entity.qualified_name  # type: ignore
         except AttributeError:
             # if we're here then it's not a cog, group, or command.
             return None
 
-        await cmd.prepare_help_command(self, entity.qualified_name)
+        await cmd.prepare_help_command(self, entity.qualified_name)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/context.py:379]: This might want to be replaced by an `is None` check?

File: discord/ext/commands/core.py [hunk 13]
```diff
@@ -70,6 +98,22 @@
...
+MISSING: Any = discord.utils.MISSING
+
+T = TypeVar('T')
+CT = TypeVar('CT', bound='Cog')
+CMT = TypeVar('CMT', bound='Command')
+CXT = TypeVar('CXT', bound='Context')
+# CHT = TypeVar('CHT', bound='Check')
+GT = TypeVar('GT', bound='Group')
+HT = TypeVar('HT', bound='Hook')
+ET = TypeVar('ET', bound='Error')
```
Comment (bijij) [inline discord/ext/commands/core.py:110]: Again with the, typevar naming convention.

I think abbreviated PascalCase With a T on the end e.g. `CtxT`, `CogT`, `GrpT` is probably the way to go?

File: discord/ext/commands/core.py [hunk 14]
```diff
@@ -160,7 +204,7 @@ def pop(self, k, default=None):
     def __setitem__(self, k, v):
         super().__setitem__(k.casefold(), v)
 
-class Command(_BaseCommand):
+class Command(_BaseCommand, Generic[CT, P, T]):
```
Comment (bijij) [inline discord/ext/commands/core.py:207]: Do we like ParamSpec?

Pros:
 - command callable arguments will be validated in-decorators.
 - `Command.__call__` / `Context.invoke` will validate arguments.

Cons:
 - code is unreadable / requires arcane knowledge to work with

File: discord/ext/commands/core.py [hunk 15]
```diff
@@ -271,74 +320,85 @@ def __init__(self, func, **kwargs):
...
 
         try:
             checks = func.__commands_checks__
             checks.reverse()
         except AttributeError:
             checks = kwargs.get('checks', [])
-        finally:
-            self.checks = checks
+
+        self.checks: List[Check] = checks
```
Comment (bijij) [inline discord/ext/commands/core.py:343]: `finally` blocks make type-checker complain about possibly unbound variables.

File: discord/ext/commands/core.py [hunk 16]
```diff
@@ -271,74 +320,85 @@ def __init__(self, func, **kwargs):
...
+
+        self.require_var_positional: bool = kwargs.get('require_var_positional', False)
+        self.ignore_extra: bool = kwargs.get('ignore_extra', True)
+        self.cooldown_after_parsing: bool = kwargs.get('cooldown_after_parsing', False)
+        self.cog: Optional[CT] = None
 
         # bandaid for the fact that sometimes parent can be the bot instance
         parent = kwargs.get('parent')
-        self.parent = parent if isinstance(parent, _BaseCommand) else None
+        self.parent: Optional[Command] = parent if isinstance(parent, _BaseCommand) else None  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/core.py:373]: Should this be `GroupMixin` or `Group`?
Technically this could be the `Bot`.

I've left as `Command` for now.

File: discord/ext/commands/core.py [hunk 17]
```diff
@@ -1157,10 +1217,10 @@ def remove_command(self, name):
             # with an already existing command. If this is the case, we want to
             # make sure the pre-existing command is not removed.
             if cmd not in (None, command):
-                self.all_commands[alias] = cmd
+                self.all_commands[alias] = cmd  # type: ignore
         return command
```
Comment (bijij) [inline discord/ext/commands/core.py:1221]: Pyright doesn't understand this guard, to remove the type ignore we'd need to change it.

File: discord/ext/commands/core.py [hunk 18]
```diff
@@ -1309,7 +1421,7 @@ async def invoke(self, ctx):
             injected = hooked_wrapped_callback(self, ctx, self.callback)
             await injected(*ctx.args, **ctx.kwargs)
 
-        ctx.invoked_parents.append(ctx.invoked_with)
+        ctx.invoked_parents.append(ctx.invoked_with)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/core.py:1424]: type checker thinks `ctx.invoked_with` could be `None` at this point.

File: discord/ext/commands/core.py [hunk 19]
```diff
@@ -1602,21 +1799,21 @@ def has_role(item):
...
+    def predicate(ctx: Context) -> bool:
         if not isinstance(ctx.channel, discord.abc.GuildChannel):
             raise NoPrivateMessage()
 
         if isinstance(item, int):
-            role = discord.utils.get(ctx.author.roles, id=item)
+            role = discord.utils.get(ctx.author.roles, id=item)  # type: ignore
         else:
-            role = discord.utils.get(ctx.author.roles, name=item)
+            role = discord.utils.get(ctx.author.roles, name=item)  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/core.py:1809]: Type-checker's going to think `ctx.author` could still be a `User`, could solve by checking the `ctx.author` instance is a `Member` instead of checking the `channel` is a `GuildChannel`

File: discord/ext/commands/core.py [hunk 20]
```diff
@@ -1799,11 +1996,11 @@ def has_guild_permissions(**perms):
     if invalid:
         raise TypeError(f"Invalid permission(s): {', '.join(invalid)}")
 
-    def predicate(ctx):
+    def predicate(ctx: Context) -> bool:
         if not ctx.guild:
             raise NoPrivateMessage
 
-        permissions = ctx.author.guild_permissions
+        permissions = ctx.author.guild_permissions  # type: ignore
```
Comment (bijij) [inline discord/ext/commands/core.py:2003]: Same issue as above with `ctx.author`

**Review by bijij — COMMENTED** [review]
Here's my initial thoughts on things that need changing.

I've ignored the Giganting overload cancer for now as removing ParamSpec deletes that right away,

And I'm going to have a crack at simplifying it now.

File: discord/ext/commands/bot.py [hunk 21]
```diff
@@ -22,13 +22,19 @@
 DEALINGS IN THE SOFTWARE.
 """
 
+from __future__ import annotations
+
+
 import asyncio
 import collections
+import collections.abc
+from discord.abc import MISSING
```
Comment (NCPlayz) [inline discord/ext/commands/bot.py:31]:
```suggestion
from discord.utils import MISSING
```

File: discord/ext/commands/context.py [hunk 22]
```diff
@@ -21,16 +21,51 @@
...
+import inspect
+import re
+
+from typing import Any, Dict, Generic, List, Optional, TYPE_CHECKING, TypeVar, Union
 
 import discord.abc
 import discord.utils
-import re
+
+from ...message import Message
```
Comment (NCPlayz) [inline discord/ext/commands/context.py:34]:
```suggestion
from discord.message import Message
```

File: discord/ext/commands/context.py [hunk 23]
```diff
@@ -21,16 +21,51 @@
...
+
+if TYPE_CHECKING:
+    from typing_extensions import ParamSpec
+
+    from ...guild import Guild
+    from ...member import Member
+    from ...state import ConnectionState
+    from ...user import ClientUser, User
+    from ...voice_client import VoiceProtocol
```
Comment (NCPlayz) [inline discord/ext/commands/context.py:43]:
```suggestion
    from discord.guild import Guild
    from discord.member import Member
    from discord.state import ConnectionState
    from discord.user import ClientUser, User
    from discord.voice_client import VoiceProtocol
```

File: discord/ext/commands/context.py [hunk 24]
```diff
@@ -248,26 +293,26 @@ def channel(self):
...
 
     @property
-    def voice_client(self):
+    def voice_client(self) -> Optional[VoiceProtocol]:
         r"""Optional[:class:`.VoiceProtocol`]: A shortcut to :attr:`.Guild.voice_client`\, if applicable."""
         g = self.guild
         return g.voice_client if g else None
 
-    async def send_help(self, *args):
+    async def send_help(self, *args: Any) -> Optional[HelpCommand]:
```
Comment (NCPlayz) [inline discord/ext/commands/context.py:315]: Split into overloads?

File: discord/ext/commands/core.py [hunk 25]
```diff
@@ -629,7 +689,7 @@ def qualified_name(self):
     def __str__(self):
```
Comment (NCPlayz) [inline discord/ext/commands/core.py:629]: Should probably type dunders for the sake of consistency.

**Review by NCPlayz — CHANGES_REQUESTED** [review]
Skimmed over changes.

File: discord/ext/commands/bot.py (hunk 1 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:151]: I believe you should be able to set `self: Client` in the method signature

File: discord/ext/commands/bot.py (hunk 5 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:66]: A convention I've seen used that I think is helpful is `CoroFuncT` and `ContextT`

File: discord/ext/commands/bot.py (hunk 1 above)
Comment (bijij) [reply discord/ext/commands/bot.py:151]: image
seems at least with pyright this is not the case.

Comment (bryanforbes) [reply discord/ext/commands/bot.py:152]: Ok, I must have misread something about advanced uses of `self` types. I'll see if I can figure out a way to do this right.

File: discord/ext/commands/cog.py [hunk 26]
```diff
@@ -89,8 +104,12 @@ async def foo(self, ctx):
                 async def bar(self, ctx):
                     pass # hidden -> False
     """
+    __cog_name__: str
+    __cog_settings__: Dict[str, Any]
+    __cog_commands__: List[Command]
+    __cog_listeners__: List[Tuple[str, str]]
```
Comment (bryanforbes) [inline discord/ext/commands/cog.py:110]: These should technically all be `ClassVar[]`

File: discord/ext/commands/cog.py (hunk 6 above)
Comment (bryanforbes) [reply discord/ext/commands/cog.py:189]: I've tried a few things and I can't get anything to work. This may be a bug in the metaclass type spec or in all of the type checkers.

File: discord/ext/commands/_types.py [hunk 27]
```diff
@@ -22,6 +22,26 @@
...
+    from .cog import Cog
+    from .errors import CommandError
+
+T = TypeVar('T')
+
+Coro = Coroutine[Any, Any, T]
+MaybeCoro = Union[T, Coro[T]]
+CoroFunc = Callable[..., Coro[Any]]
+
+Check = Union[Callable[["Cog", "Context"], MaybeCoro[bool]], Callable[["Context"], MaybeCoro[bool]]]
```
Comment (bryanforbes) [inline discord/ext/commands/_types.py:40]: Should `"Context"` be `"Context[Any]"`?

File: discord/ext/commands/context.py (hunk 9 above)
Comment (bryanforbes) [reply discord/ext/commands/context.py:144]: Yes

File: discord/ext/commands/context.py [hunk 28]
```diff
@@ -206,20 +248,23 @@ async def reinvoke(self, *, call_hooks: bool = False, restart: bool = True):
...
 
-    async def _get_channel(self):
+    async def _get_channel(self) -> discord.abc.Messageable:
         return self.channel
 
     @property
-    def clean_prefix(self):
-        """:class:`str`: The cleaned up invoke prefix. i.e. mentions are ``@name`` instead of ``<@id>``.
+    def clean_prefix(self) -> str:
+        """Optional[:class:`str`]: The cleaned up invoke prefix. i.e. mentions are ``@name`` instead of ``<@id>``.
```
Comment (bryanforbes) [inline discord/ext/commands/context.py:260]: This documentation change doesn't match the return type

File: discord/ext/commands/core.py (hunk 16 above)
Comment (bryanforbes) [reply discord/ext/commands/core.py:373]: It should probably be `GroupMixin`

File: discord/ext/commands/core.py [hunk 29]
```diff
@@ -508,9 +568,9 @@ async def transform(self, ctx, param):
             argument = view.get_quoted_word()
         view.previous = previous
 
-        return await run_converters(ctx, converter, argument, param)
+        return await run_converters(ctx, converter, argument, param)  # type: ignore
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:571]: Why the `# type: ignore` here?

File: discord/ext/commands/core.py [hunk 30]
```diff
@@ -704,14 +764,14 @@ async def call_before_hooks(self, ctx):
...
             await hook(ctx)
 
-    async def call_after_hooks(self, ctx):
+    async def call_after_hooks(self, ctx: Context) -> None:
         cog = self.cog
         if self._after_invoke is not None:
             instance = getattr(self._after_invoke, '__self__', cog)
             if instance:
-                    await self._after_invoke(instance, ctx)
+                    await self._after_invoke(instance, ctx)  # type: ignore
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:772]: Why is this being ignored?

File: discord/ext/commands/core.py [hunk 31]
```diff
@@ -896,10 +956,10 @@ def before_invoke(self, coro):
         if not asyncio.iscoroutinefunction(coro):
             raise TypeError('The pre-invoke hook must be a coroutine.')
 
-        self._before_invoke = coro
+        self._before_invoke = coro  # type: ignore
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:959]: Why is this being ignored?

File: discord/ext/commands/core.py [hunk 32]
```diff
@@ -862,17 +922,17 @@ def error(self, coro):
         if not asyncio.iscoroutinefunction(coro):
             raise TypeError('The error handler must be a coroutine.')
 
-        self.on_error = coro
+        self.on_error: Error = coro  # type: ignore
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:925]: Why is this being ignored?

File: discord/ext/commands/core.py [hunk 33]
```diff
@@ -1216,7 +1276,33 @@ def get_command(self, name):
 
         return obj
 
-    def command(self, *args, **kwargs):
+    @overload
+    def command(
+        self,
+        name: str = ...,
+        cls: Type[Command[CT, P, T]] = ...,
+        *args: Any,
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:1284]: Should these be expanded?

File: discord/ext/commands/core.py [hunk 34]
```diff
@@ -1225,15 +1311,41 @@ def command(self, *args, **kwargs):
...
 
         return decorator
 
-    def group(self, *args, **kwargs):
+    @overload
+    def group(
+        self,
+        name: str = ...,
+        cls: Type[Group[CT, P, T]] = ...,
+        *args: Any,
```
Comment (bryanforbes) [inline discord/ext/commands/core.py:1327]: Should these be expanded?

File: discord/ext/commands/core.py (hunk 14 above)
Comment (bryanforbes) [reply discord/ext/commands/core.py:207]: I think the code is pretty well readable and with better names for `TypeVar`s, the code will be come much more readable

**Review by bryanforbes — CHANGES_REQUESTED** [review]
Overall this looks really good. One recommendation I have is that anywhere you have a `# type: ignore`, make sure to leave a comment as to why it's being ignored. It's not always clear.

File: discord/ext/commands/core.py (hunk 32 above)
Comment (bijij) [reply discord/ext/commands/core.py:925]: No clue

File: discord/ext/commands/core.py (hunk 30 above)
Comment (bijij) [reply discord/ext/commands/core.py:772]: type-checker believes argument could still be none, would need to cast.

File: discord/ext/commands/core.py (hunk 29 above)
Comment (bijij) [reply discord/ext/commands/core.py:571]: type-checker believes argument could still be none, would need to cast.

File: discord/ext/commands/context.py (hunk 24 above)
Comment (bijij) [reply discord/ext/commands/context.py]: I left as Any, I feel like, given helps probably getting a rework might not be worth the effort yet.

File: discord/ext/commands/core.py (hunk 17 above)
Comment (Rapptz) [reply discord/ext/commands/core.py:1219]: Does it understand `if cmd is not None and cmd != command`?

File: discord/ext/commands/errors.py [hunk 35]
```diff
@@ -971,4 +971,4 @@ class MissingFlagArgument(FlagError):
     """
     def __init__(self, flag):
         self.flag = flag
-        super().__init__(f'Flag {flag.name!r} does not have an argument')
+        super().__init__(f'Flag {flag.name!r} does not have an argument')
```
Comment (Rapptz) [inline discord/ext/commands/errors.py:974]: Why did you remove the newline at the end of this file?

File: discord/ext/commands/bot.py [hunk 36]
```diff
@@ -200,10 +220,10 @@ def check_commands(ctx):
                 return ctx.command.qualified_name in allowed_commands
 
         """
-        self.add_check(func)
+        self.add_check(func)  # type: ignore
```
[the remark this answers is not in the document]
Comment (bijij) [reply discord/ext/commands/bot.py:223]: T has a different bound to add_check so it won't type check. would need HKT or to lose the type on the return.

File: discord/ext/commands/errors.py (hunk 35 above)
Comment (bijij) [reply discord/ext/commands/errors.py:974]: Because I'm bad.

File: discord/ext/commands/context.py (hunk 12 above)
Comment (bijij) [reply discord/ext/commands/context.py]: Type-checker reset scope thinking entity could still be None. resolved.

File: discord/ext/commands/core.py [hunk 37]
```diff
@@ -384,16 +442,16 @@ def remove_check(self, func):
         except ValueError:
             pass
 
-    def update(self, **kwargs):
-        """Updates :class:`Command` instance with updated attributes.
+    def update(self, **kwargs: Any) -> None:
+        """Updates :class:`Command` instance with updated attribute.
```
Comment (Gobot1234) [inline discord/ext/commands/core.py:446]: This seems like a typo

File: discord/ext/commands/help.py [hunk 38]
```diff
@@ -320,7 +326,7 @@ def __init__(self, **options):
         self.command_attrs = attrs = options.pop('command_attrs', {})
         attrs.setdefault('name', 'help')
         attrs.setdefault('help', 'Shows this message')
-        self.context = None
+        self.context: Optional[Context] = None
         self._command_impl = _HelpCommandImpl(self, **self.command_attrs)
 
     def copy(self):
```
Comment (Gobot1234) [inline discord/ext/commands/help.py:332]: This is untyped

**Review by Rapptz — APPROVED** [review]
LGTM. Thanks.
